toolchain/lex/tokenized_buffer.h

```diff
@@ -260,32 +260,93 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
 
-  struct TokenInfo {
-    TokenKind kind;
+  class TokenInfo {
+   public:
+    auto kind() const -> TokenKind { return TokenKind::FromInt(kind_); }
+    auto has_leading_space() const -> bool { return has_leading_space_; }
+
+    auto ident_id() const -> IdentifierId {
+      CARBON_DCHECK(kind() == TokenKind::Identifier);
+      return IdentifierId(token_payload_);
+    }
+    auto set_ident_id(IdentifierId ident_id) -> void {
+      CARBON_DCHECK(kind() == TokenKind::Identifier);
+      CARBON_DCHECK(ident_id.index < (2 << PayloadBits));
+      token_payload_ = ident_id.index;
+    }
+
+    auto string_literal_id() const -> StringLiteralValueId {
+      CARBON_DCHECK(kind() == TokenKind::StringLiteral);
+      return StringLiteralValueId(token_payload_);
+    }
+
+    auto int_id() const -> IntId {
+      CARBON_DCHECK(kind() == TokenKind::IntLiteral ||
+                    kind() == TokenKind::IntTypeLiteral ||
+                    kind() == TokenKind::UnsignedIntTypeLiteral ||
+                    kind() == TokenKind::FloatTypeLiteral);
+      return IntId(token_payload_);
+    }
+
+    auto real_id() const -> RealId {
+      CARBON_DCHECK(kind() == TokenKind::RealLiteral);
+      return RealId(token_payload_);
+    }
+
+    auto closing_token_index() const -> TokenIndex {
+      CARBON_DCHECK(kind().is_opening_symbol());
+      return TokenIndex(token_payload_);
+    }
+    auto set_closing_token_index(TokenIndex closing_index) -> void {
+      CARBON_DCHECK(kind().is_opening_symbol());
+      CARBON_DCHECK(closing_index.index < (2 << PayloadBits));
+      token_payload_ = closing_index.index;
+    }
+
+    auto opening_token_index() const -> TokenIndex {
+      CARBON_DCHECK(kind().is_closing_symbol());
+      return TokenIndex(token_payload_);
+    }
+    auto set_opening_token_index(TokenIndex opening_index) -> void {
+      CARBON_DCHECK(kind().is_closing_symbol());
+      CARBON_DCHECK(opening_index.index < (2 << PayloadBits));
+      token_payload_ = opening_index.index;
+    }
+
+    auto error_length() const -> int {
+      CARBON_DCHECK(kind() == TokenKind::Error);
+      return token_payload_;
+    }
+
+    auto byte_offset() const -> int32_t { return byte_offset_; }
+
+    TokenInfo(TokenKind kind, bool has_leading_space, int32_t byte_offset)
+        : kind_(kind.AsInt()),
+          has_leading_space_(has_leading_space),
+          byte_offset_(byte_offset) {}
+
+    TokenInfo(TokenKind kind, bool has_leading_space, int payload,
+              int32_t byte_offset)
+        : kind_(kind.AsInt()),
+          has_leading_space_(has_leading_space),
+          token_payload_(payload),
+          byte_offset_(byte_offset) {
+      CARBON_DCHECK(payload >= 0 && payload < (2 << PayloadBits));
+    }
+
+   private:
+    static constexpr int PayloadBits = 23;
+
+    unsigned kind_ : sizeof(TokenKind) * 8;
 
     // Whether the token has trailing whitespace.
-    bool has_trailing_space = false;
+    unsigned has_leading_space_ : 1;
```

Comment:
Maybe recent versions of C++ or C++ compilers have fixed this, but I have an old habit of making bitfields unsigned to avoid any implementation oddities around sign extension. My memory is that `bool` specifically broke things in LLVM when used in a bitfield.

The thing I found most uniformly consistent in behavior was to declare *everything* as `unsigned` and manually manage the bits going in and out and then transform them into proper types.

If this is no longer an issue, happy to adjust though. @zygoloid might recall the specific issue that caused us to start following this practice in LLVM code.

Reply:
Looking around, in [subst.cpp](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/subst.cpp#L21-L27) zygoloid added typed bit-packed values, so agreed it'd be good to figure out.

Note, I would still expect the accessors because the `RawEnumType` is unwieldy to work with directly.

Reply:
Hmm, ok, that seems reasonable evidence that this works well enough these days. Switched to use slightly better types, and removed the added behavior in token kind that we don't need now. =] Nice!

I kept `unsigned` on the payload because the plan is to use the extra bit and not store negative payload indices, so it seems a reasonable type.

(also, sorry for losing this comment in the discussion)